### DISCUSSION TITLE
Getting a "AttributeError: module 'hashlib' has no attribute 'scrypt'" during the tutorial

### DISCUSSION THREAD

**Discussion by SethBashamDev** [pallets/flask#5325] [Q&A]
My code matches 1 for 1; followed the tutorial to the T and I am getting the error posted in the title. Can't seem to get an answer anywhere.

**Comment by davidism** [comment]
What version of Python are you using and where did you install it from? scrypt is available in all supported builds of Python.

https://docs.python.org/3/library/hashlib.html#hashlib.scrypt note the docs say it was added in 3.6.

**Reply by SethBashamDev** [reply]
I am using 3.9.6. I even tried downloading the completed Flaskr app and registering a user and it gave me the same results.

**Reply by jonnolen** [reply]
i'm getting the same thing using python 3.9.6 on macos montery 12.7.2 (intel mac)

**Reply by Frazier-Stewart** [reply]
same problem on apple chip

**Reply by jonnolen** [reply]
i had to install python3 with homebrew, recreate my venv and then i  no longer got the error.

**Reply by davidism** [reply]
I can't reproduce this issue with any supported version of Python on Mac. Python installed both from https://python.org and homebrew works as expected. It sounds like there's another issue with your setup that is not related to Flask.

**Reply by jonnolen** [reply]
it seems to only happen with the system provided python3.

**Reply by Frazier-Stewart** [reply]
Nice! I installed python3.10 via brew, and the problem is solved.

**Reply by davidism** [reply]
Note that both the system and brew Python's are not intended for "normal" development. You'd only develop against them if you expect to install next to and use them. Typically, you should use the official installers available at https://python.org.

**Reply by xwtx66666666** [reply]
I'm studying Flask and i create a Flask project named pythonbbs. In define UserModel, i imported the function generate_password_hash from package werkzeug.security for encryption the password from front html. In Windows, it runs well. but when i pull codes to centos 7.9, i also got this error. My python version is 3.9.0 in linux but 3.9.16 in windows. What need i do to resolve this problem?

**Reply by ThiefMaster** [reply]
start by not using an ancient python version

**Reply by scepticore** [reply]
> start by not using an ancient python version

Edit 0: My Versions
Python 3.12.4
OpenSSL OpenSSL 3.3.1
macOS Sonoma 14.5

I have the same issue, but I'm using Python 3.12 and I just freshly installed the tutorial folder by cloning it from github with the tutorial-steps given. Is there anything else I might check? Any other package seems to be up to date as well.

Edit 1:
Another thing I tried: I imported werkzeug into a separate scratch file and tested the generate_password_hash(password) and it worked fine. It's just within the Flaskr folder, where I get this AttributeError.

Edit 2:
I just updated LibreSSL as well as OpenSSL to it's latest version, the issue still appears.

**Comment by easyop** [comment]
我不确定这是不是Werkzeug的某个版本导致的问题。 generate_password_hash(password)的 default method变成了scrypt。 我是在调用generate_password_hash(password,method='pbkdf2')时显示的制定method为pbkdf2来解决的此问题。
